lib/srv/desktop/rdp/rdpclient/src/client.rs

```diff
@@ -14,129 +13,327 @@ use ironrdp_pdu::rdp::RdpError;
-        None,
-    );
-
-    loop {
-        tokio::select! {
-             res = upgraded_framed.read_pdu() => {
-                let (action, mut frame) = res?;
-                match action {
-                    ironrdp_pdu::Action::X224 => {
-                        let result =
-                            process_x224_frame(&mut x224_processor, &frame).await;
-                        process_active_stage_result(&mut upgraded_framed, result).await.map_err(|e|{
-                                error!("process_stage_result {:?}", e);
-                                ConnectError::Rdp(RdpError::InvalidSecurityHeader)
-                            })?;
-                    },
-                    ironrdp_pdu::Action::FastPath => {
-                            frame_sender.send(frame).await;
-                    },
-                };
-             }
-             Some(data) = tdp_reader.recv() => {
-                match data {
-                    TdpMessage::Key(p) => {
-                        write_rdp_key(&mut upgraded_framed, p).await;
+    /// Spawns a task for running the RDP loop which:
+    /// 1. Reads new frames from the RDP server and sends them to Go.
+    /// 2. Listens on the Client's function_receiver for function calls
+    ///    which it then executes.
+    ///
+    /// Returns immediately with a receiver which callers are expected to listen
+    /// on in case of any errors, or until a [`ClientFunction::Stop`] is received.
+    ///
+    /// The caller is responsible for ensuring that the future spawned by this function
+    /// eventually returns. Failure to do so can result in a leak.
+    fn run_rdp_loop(self) -> oneshot::Receiver<Result<(), ClientError>> {
+        let (result_tx, result_rx) = oneshot::channel::<Result<(), ClientError>>();
+
+        global::TOKIO_RT.spawn(async move {
+            let res = self.run_rdp_loop_internal().await;
+            match result_tx.send(res) {
+                Ok(_) => {}
+                Err(res) => {
+                    error!("failed to send result: {:?}", res)
+                }
+            };
+        });
+
+        result_rx
+    }
+
+    async fn run_rdp_loop_internal(mut self) -> Result<(), ClientError> {
+        if let Some(mut function_receiver) = self.function_receiver.take() {
+            loop {
+                tokio::select! {
+                     res = self.rdp_stream.read_pdu() => {
+                        let (action, mut frame) = res?;
+                        match action {
+                            ironrdp_pdu::Action::X224 => {
+                                let result = self.process_x224_frame(&frame).await;
+                                self.process_active_stage_result(result).await?;
+                            },
+                            ironrdp_pdu::Action::FastPath => {
+                                unsafe {
+                                    handle_remote_fx_frame(self.cgo_handle, frame.as_mut_ptr(), frame.len() as u32);
```

Comment:
Calling `handle_remote_fx_frame` inside select is problematic as this method will block. I'd suggest to either send it through channel to caller thread as I did before or wrap it with `spawn_blocking` and `await` on `JoinHandle` returned.
It might be a good idea to refactor in a way that we can handle user input while we're waiting on sending frame

Reply:
This is a good point, it's not something that we had to consider previously when the rust side was all synchronous.

Another option is that we write all of our functions in Go to be non-blocking, so `handle_remote_fx_frame` becomes

```go
//export handle_remote_fx_frame
func handle_remote_fx_frame(handle C.uintptr_t, data *C.uint8_t, length C.uint32_t) {
	go func() {
		goData := asRustBackedSlice(data, int(length))
		cgo.Handle(handle).Value().(*Client).handleRDPFastPathPDU(goData)
	}()
}
```

I believe this is functionally equivalent to `spawn_blocking` without `await`ing on the `JoinHandle`.

In fact I think all of these options are in effect functionally equivalent to blocking in another, blocking thread and continuing without waiting for the blocking call to complete:

- send it through channel to caller thread
- wrapping in a goroutine on the go side
- `spawn_blocking` without `await`ing on the `JoinHandle`

If we do need to wait for the operation to complete then we can safely do

- `spawn_blocking` and `await` on `JoinHandle` returned

but as you point out this is less than ideal in our primary control loop.

Given that understanding I think we should go with [`spawn_blocking` without `await`ing on the `JoinHandle`] here and wherever possible, and [`spawn_blocking` and `await` on `JoinHandle` returned] whenever the ordering of calls is important.

Reply:
`go func` will not work as you can't hold on to data pointer past the return of the `handle_remote_fx_frame`.
And without waiting for `handle_remote_fx_frame` to finish we risk that frames will arrive out of order (that was not the case with channel as we receive data in the same order we put it in.
There is a mechanism in `tokio::select!` that let you disable branches based on conditions, we could probably use that to allow inputs during frame handling

Reply:
> go func will not work as you can't hold on to data pointer past the return of the handle_remote_fx_frame

True.

> And without waiting for handle_remote_fx_frame to finish we risk that frames will arrive out of order (that was not the case with channel as we receive data in the same order we put it in.

Also true.

> There is a mechanism in tokio::select! that let you disable branches based on conditions, we could probably use that to allow inputs during frame handling

I'm not quite sure how that would work... Btw just to be clear, your original idea of sending via a channel back to the parent blocking thread does allow inputs during frame handling.